rust/kernel/types.rs

```diff
@@ -200,29 +206,66 @@ impl<T: PointerWrapper + Deref> PointerWrapper for Pin<T> {
 ///     pr_info!("example2 no early return\n");
 /// }
 /// ```
-pub struct ScopeGuard<T: FnOnce()> {
-    cleanup_func: Option<T>,
+///
+/// In the example below, we need a mutable object (the vector) to be accessible within the log
+/// function, so we wrap it in the [`ScopeGuard`]:
+/// ```
+/// # use kernel::prelude::*;
+/// # use kernel::ScopeGuard;
+/// fn example3(arg: bool) -> Result {
+///     let mut vec = ScopeGuard::new(Vec::new(), |v| pr_info!("vec had {} elements\n", v.len()));
+///
+///     vec.try_push(10u8)?;
+///     if arg {
+///         return Ok(());
+///     }
+///     vec.try_push(20u8)?;
+///     Ok(())
+/// }
+/// ```
+pub struct ScopeGuard<T, F: FnOnce(&mut T)> {
+    cleanup_func: Option<F>,
+    data: T,
 }
 
-impl<T: FnOnce()> ScopeGuard<T> {
-    /// Creates a new cleanup object with the given cleanup function.
-    pub fn new(cleanup_func: T) -> Self {
+impl<T, F: FnOnce(&mut T)> ScopeGuard<T, F> {
+    /// Creates a new guarded object with the given cleanup function.
+    pub fn new(data: T, cleanup_func: F) -> Self {
         Self {
+            data,
             cleanup_func: Some(cleanup_func),
         }
     }
 
-    /// Prevents the cleanup function from running.
-    pub fn dismiss(mut self) {
+    /// Prevents the cleanup function from running and returns the guarded data.
+    pub fn dismiss(mut self) -> T {
         self.cleanup_func.take();
+        // SAFETY: `data` is valid for reads, properly aligned, and contains initialised data.
+        let data = unsafe { core::ptr::read(&self.data) };
+        core::mem::forget(self);
+        data
+    }
+}
+
+impl<T, F: FnOnce(&mut T)> Deref for ScopeGuard<T, F> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.data
+    }
+}
+
+impl<T, F: FnOnce(&mut T)> DerefMut for ScopeGuard<T, F> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.data
     }
 }
 
-impl<T: FnOnce()> Drop for ScopeGuard<T> {
+impl<T, F: FnOnce(&mut T)> Drop for ScopeGuard<T, F> {
```

Comment:
Can the closure consume the data instead?

Reply:
Do you have a case in mind where this would be useful?

It can certainly be done, but I'd have to add another unsafe block, which I'd rather not do unless there is an actual need for it. (If the need arises in the future, we could add it then.)

Reply:
I think I definitely see a potential use of scope guard that needs to take an ownership.

API design wise, it doesn't make much sense for the function to take `&mut T` since `T` is going to be dropped anyway. Ergnomics IMO is more important than number of unsafe blocks in a basic shared utility (plus you have already used unsafe in dismiss anyway).

Reply:
You can also use `Option<F, T>` to get rid of any unsafe.

Reply:
@bjorn3 @nbdd0121 I implemented a version without unsafe using `Option<T, F>`. The problem is that the optimiser does a horrible job with it. For example, consider the following trivial example:

```rust
extern "C" {
    fn testfn(x: u8);
}

pub fn example3(arg: bool, a: u8) -> u8 {
     let mut v = ScopeGuard::new(10u8, |v| unsafe { testfn(*v) });

     *v += 1;
     if arg {
         return *v;
     }
     v.dismiss() + a
}
```

The result is of good quality ([code here](https://rust.godbolt.org/z/Yx36f3z16)) with the `ScopeGuard` in this PR (I don't quite get why we need `pushq %rax` and the subsequent `addq $8, %rsp` but I digress. Actually, isn't this breaking the 16-byte alignment requirement of the C calling convention on x86-64?)

```
example::example3:
        testl   %edi, %edi
        je      .LBB0_3
        pushq   %rax
        movl    $11, %edi
        callq   *testfn@GOTPCREL(%rip)
        movb    $11, %al
        addq    $8, %rsp
        retq
.LBB0_3:
        movl    %esi, %eax
        addb    $11, %al
        retq
```

Now the `Option`-based implementation is completely butchered ([code here](https://rust.godbolt.org/z/eojr91e61)):

```
example::example3:
        pushq   %rbp
        pushq   %rbx
        pushq   %rax
        testl   %edi, %edi
        je      .LBB1_2
        movl    $11, %edi
        callq   *testfn@GOTPCREL(%rip)
        movb    $11, %al
        addq    $8, %rsp
        popq    %rbx
        popq    %rbp
        retq
.LBB1_2:
        movw    $2817, 6(%rsp)
        movb    $0, 6(%rsp)
        movb    $1, %al
        testb   %al, %al
        je      .LBB1_7
        movl    %esi, %ebx
        movw    $2817, %bp
        movzwl  6(%rsp), %eax
        testb   $1, %al
        je      .LBB1_5
        movzwl  %ax, %edi
        shrl    $8, %edi
        callq   *testfn@GOTPCREL(%rip)
.LBB1_5:
        movzwl  %bp, %eax
        shrl    $8, %eax
        addb    %bl, %al
        addq    $8, %rsp
        popq    %rbx
        popq    %rbp
        retq
.LBB1_7:
        leaq    .L__unnamed_1(%rip), %rdi
        leaq    .L__unnamed_2(%rip), %rdx
        movl    $43, %esi
        callq   *core::panicking::panic@GOTPCREL(%rip)
        ud2
        movq    %rax, %rbx
        leaq    6(%rsp), %rdi
        callq   core::ptr::drop_in_place<example::ScopeGuard<u8,example::example3::{{closure}}>>
        movq    %rbx, %rdi
        callq   _Unwind_Resume@PLT
        ud2
```

Note that we have sequences that seem trivially optimisable, for example:

```
        movb    $1, %al
        testb   %al, %al
        je      .LBB1_7
```

Do you know what could be going on here?

Reply:
The conditional in `unwrap()` in `dismiss()` is tripping it, with `unsafe { self.0.take().unwrap_unchecked().0 }` it optimizes to the same.

Reply:
Thanks Miguel! I guess we could use this workaround for now.

But I'm still interested in hearing about _why_ the compiler is failing to optimise with `.unwrap`.

Reply:
It seems to be related to unwinding path. If you add `-C panic=abort` then it optimises fine.

Reply:
> Actually, isn't this breaking the 16-byte alignment requirement of the C calling convention on x86-64?

No, before the call instruction it needs to be 16 byte aligned. When this function was called, the return address was pushed on the stack, causing the stack to get aligned to 8 bytes, but not 16 bytes. Another push is necessary to align it to 16 bytes again.

Reply:
> It seems to be related to unwinding path. If you add `-C panic=abort` then it optimises fine.

Indeed. Since we are compiling with `-C panic=abort`, I checked the code in the kernel (with a slight modification to use `wrapping_add`) and we do get the better version:

```
ffffffff8181f220 <_RNvNtCshhMfd4mY5QU_6kernel5types8example3>:
ffffffff8181f220:       85 ff                   test   %edi,%edi
ffffffff8181f222:       74 12                   je     ffffffff8181f236 <_RNvNtCshhMfd4mY5QU_6kernel5types8example3+0x16>
ffffffff8181f224:       55                      push   %rbp
ffffffff8181f225:       48 89 e5                mov    %rsp,%rbp
ffffffff8181f228:       bf 0b 00 00 00          mov    $0xb,%edi
ffffffff8181f22d:       e8 7e cd ff ff          call   ffffffff8181bfb0 <testfn>
ffffffff8181f232:       b0 0b                   mov    $0xb,%al
ffffffff8181f234:       5d                      pop    %rbp
ffffffff8181f235:       c3                      ret
ffffffff8181f236:       89 f0                   mov    %esi,%eax
ffffffff8181f238:       04 0b                   add    $0xb,%al
ffffffff8181f23a:       c3                      ret
```

Reply:
> > Actually, isn't this breaking the 16-byte alignment requirement of the C calling convention on x86-64?
> 
> No, before the call instruction it needs to be 16 byte aligned. When this function was called, the return address was pushed on the stack, causing the stack to get aligned to 8 bytes, but not 16 bytes. Another push is necessary to align it to 16 bytes again.

Ah, thanks for the clarification! I had grown accustomed to aarch64's storing of the return pointer in the link register, so no changes to the callstack on the branching instruction.